In [ ]:
#Import library
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [ ]:
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


In [ ]:
#load dataset
df = pd.read_csv('/content/gdrive/MyDrive/Colab Notebooks/hotel_bookings_data.csv')
pd.set_option('display.max_columns', None) #setting to view all columns
print(df.shape) #see the number of rows and columns
df.sample(3) #displays sample data of 3

(119390, 29)


hotel  is_canceled  lead_time  arrival_date_year  \
25586  Resort Hotel            0        167               2018   
42025    City Hotel            0         12               2017   
87994    City Hotel            0         67               2018   

      arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
25586             August                        27                         26   
42025            October                        36                         30   
87994               June                        18                         26   

       stays_in_weekend_nights  stays_in_weekdays_nights  adults  children  \
25586                        2                         5       2       0.0   
42025                        1                         0       2       0.0   
87994                        0                         3       3       0.0   

       babies       meal           city market_segment distribution_channel  \
25586       0  Breakfast    Kota Malang      Online TA                TA/TO   
42025       0  Breakfast  Kota Denpasar         Direct               Direct   
87994       0  Breakfast  Kota Bengkulu         Direct               Direct   

       is_repeated_guest  previous_cancellations  \
25586                  0                       0   
42025                  0                       0   
87994                  0                       0   

       previous_bookings_not_canceled  booking_changes deposit_type  agent  \
25586                               0                0   No Deposit  240.0   
42025                               0                0   No Deposit   14.0   
87994                               0                0   No Deposit   14.0   

       company  days_in_waiting_list customer_type     adr  \
25586      NaN                     0        Family   76.57   
42025      NaN                     0      Personal   79.00   
87994      NaN                     0      Personal  121.10   

       required_car_parking_spaces  total_of_special_requests  \
25586                            0                          0   
42025                            0                          0   
87994                            0                          1   

      reservation_status  
25586          Check-Out  
42025          Check-Out  
87994          Check-Out

# Data Preprocessing

## Information of DataFrame

In [ ]:
#summary of a DataFrame
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 29 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_weekdays_nights        119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

The data consists of 119,390 rows and 29 columns and no changes were made to the data type.

In [ ]:
df.describe() #Generate descriptive statistics

is_canceled      lead_time  arrival_date_year  \
count  119390.000000  119390.000000      119390.000000   
mean        0.370416     104.011416        2018.156554   
std         0.482918     106.863097           0.707476   
min         0.000000       0.000000        2017.000000   
25%         0.000000      18.000000        2018.000000   
50%         0.000000      69.000000        2018.000000   
75%         1.000000     160.000000        2019.000000   
max         1.000000     737.000000        2019.000000   

       arrival_date_week_number  arrival_date_day_of_month  \
count             119390.000000              119390.000000   
mean                  27.165173                  15.798241   
std                   13.605138                   8.780829   
min                    1.000000                   1.000000   
25%                   16.000000                   8.000000   
50%                   28.000000                  16.000000   
75%                   38.000000                  23.000000   
max                   53.000000                  31.000000   

       stays_in_weekend_nights  stays_in_weekdays_nights         adults  \
count            119390.000000             119390.000000  119390.000000   
mean                  0.927599                  2.500302       1.856403   
std                   0.998613                  1.908286       0.579261   
min                   0.000000                  0.000000       0.000000   
25%                   0.000000                  1.000000       2.000000   
50%                   1.000000                  2.000000       2.000000   
75%                   2.000000                  3.000000       2.000000   
max                  19.000000                 50.000000      55.000000   

            children         babies  is_repeated_guest  \
count  119386.000000  119390.000000      119390.000000   
mean        0.103890       0.007949           0.031912   
std         0.398561       0.097436           0.175767   
min         0.000000       0.000000           0.000000   
25%         0.000000       0.000000           0.000000   
50%         0.000000       0.000000           0.000000   
75%         0.000000       0.000000           0.000000   
max        10.000000      10.000000           1.000000   

       previous_cancellations  previous_bookings_not_canceled  \
count           119390.000000                   119390.000000   
mean                 0.087118                        0.137097   
std                  0.844336                        1.497437   
min                  0.000000                        0.000000   
25%                  0.000000                        0.000000   
50%                  0.000000                        0.000000   
75%                  0.000000                        0.000000   
max                 26.000000                       72.000000   

       booking_changes          agent      company  days_in_waiting_list  \
count    119390.000000  103050.000000  6797.000000         119390.000000   
mean          0.221124      86.693382   189.266735              2.321149   
std           0.652306     110.774548   131.655015             17.594721   
min           0.000000       1.000000     6.000000              0.000000   
25%           0.000000       9.000000    62.000000              0.000000   
50%           0.000000      14.000000   179.000000              0.000000   
75%           0.000000     229.000000   270.000000              0.000000   
max          21.000000     535.000000   543.000000            391.000000   

                 adr  required_car_parking_spaces  total_of_special_requests  
count  119390.000000                119390.000000              119390.000000  
mean      101.831122                     0.062518                   0.571363  
std        50.535790                     0.245291                   0.792798  
min        -6.380000                     0.000000                   0.000000  
25%        69.290000                     0.000000             

## Copy Dataframe

In [ ]:
df_clean = df.copy() #copy dataframe for Cleaning

## Missing Value

In [ ]:
#detect missing values
df_clean.isnull().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_weekdays_nights               0
adults                                 0
children                               4
babies                                 0
meal                                   0
city                                 488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
booking_changes                        0
deposit_type                           0
agent                              16340
company                           112593
days_in_waiting_list                   0
customer_type   

There are 4 columns with missing data : children, city, agent, company.

### Checking missing values

In [ ]:
#cheking missing values for children
df_clean.loc[df_clean['children'].isnull(), : ]

hotel  is_canceled  lead_time  arrival_date_year  \
40600  City Hotel            1          2               2017   
40667  City Hotel            1          1               2017   
40679  City Hotel            1          1               2017   
41160  City Hotel            1          8               2017   

      arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
40600            October                        32                          3   
40667            October                        32                          5   
40679            October                        32                          5   
41160            October                        33                         13   

       stays_in_weekend_nights  stays_in_weekdays_nights  adults  children  \
40600                        1                         0       2       NaN   
40667                        0                         2       2       NaN   
40679                        0                         2       3       NaN   
41160                        2                         5       2       NaN   

       babies       meal           city market_segment distribution_channel  \
40600       0  Breakfast  Kota Denpasar      Undefined            Undefined   
40667       0  Breakfast  Kota Denpasar         Direct            Undefined   
40679       0  Breakfast  Kota Denpasar      Undefined            Undefined   
41160       0  Breakfast  Kota Denpasar      Online TA            Undefined   

       is_repeated_guest  previous_cancellations  \
40600                  0                       0   
40667                  0                       0   
40679                  0                       0   
41160                  0                       0   

       previous_bookings_not_canceled  booking_changes deposit_type  agent  \
40600                               0                0   No Deposit    NaN   
40667                               0                0   No Deposit   14.0   
40679                               0                0   No Deposit    NaN   
41160                               0                0   No Deposit    9.0   

       company  days_in_waiting_list customer_type   adr  \
40600      NaN                     0        Family  12.0   
40667      NaN                     0        Family  12.0   
40679      NaN                     0        Family  18.0   
41160      NaN                     0        Family  76.5   

       required_car_parking_spaces  total_of_special_requests  \
40600                            0                          1   
40667                            0                          1   
40679                            0                          2   
41160                            0                          1   

      reservation_status  
40600           Canceled  
40667           Canceled  
40679           Canceled  
41160           Canceled

In [ ]:
#Cheking missing values for city
df_clean.loc[df_clean['city'].isnull(), : ]

hotel  is_canceled  lead_time  arrival_date_year  \
30      Resort Hotel            0        118               2017   
4127    Resort Hotel            1          0               2018   
7092    Resort Hotel            1          8               2018   
7860    Resort Hotel            1         39               2018   
8779    Resort Hotel            1          0               2018   
...              ...          ...        ...                ...   
65908     City Hotel            1          0               2019   
65909     City Hotel            1          0               2019   
65910     City Hotel            1          0               2019   
80830     City Hotel            0          4               2017   
101488    City Hotel            0          1               2018   

       arrival_date_month  arrival_date_week_number  \
30              September                        27   
4127                April                         8   
7092            September                        30   
7860              October                        36   
8779             December                        42   
...                   ...                       ...   
65908                June                        15   
65909                June                        15   
65910                June                        15   
80830             January                        48   
101488            January                        47   

        arrival_date_day_of_month  stays_in_weekend_nights  \
30                              1                        4   
4127                           15                        0   
7092                           21                        0   
7860                           30                        0   
8779                           13                        0   
...                           ...                      ...   
65908                          10                        0   
65909                          10                        0   
65910                          10                        0   
80830                          23                        1   
101488                         13                        2   

        stays_in_weekdays_nights  adults  children  babies       meal city  \
30                            10       1       0.0       0  Breakfast  NaN   
4127                           0       0       0.0       0    No Meal  NaN   
7092                           1       1       0.0       0  Breakfast  NaN   
7860                           5       2       0.0       0     Dinner  NaN   
8779                           1       1       0.0       0  Breakfast  NaN   
...                          ...     ...       ...     ...        ...  ...   
65908                          0       0       0.0       0    No Meal  NaN   
65909                          0       0       0.0       0    No Meal  NaN   
65910                          0       0       0.0       0    No Meal  NaN   
80830                          2       1       0.0       0  Breakfast  NaN   
101488                         2       2       0.0       0  Breakfast  NaN   

       market_segment distribution_channel  is_repeated_guest  \
30             Direct               Direct                  0   
4127    Offline TA/TO                TA/TO                  0   
7092        Corporate            Corporate                  0   
7860           Direct               Direct                  0   
8779        Corporate            Corporate                  0   
...               ...                  ...                ...   
65908   Complementary            Corporate                  0   
65909   Complementary            Corporate                  0   
65910   Complementary            Corporate                  0   
80830          Groups                TA/TO                  0   
101488         Direct               Direct                  0   

        previous_cancellations  previous_bookings_not_canceled  \
30                  

In [ ]:
#cheking missing values for agent and company
df_clean[df_clean['agent'].notnull() & df_clean['company'].notnull()]

hotel  is_canceled  lead_time  arrival_date_year  \
2392    Resort Hotel            0          6               2017   
2697    Resort Hotel            0         24               2017   
2867    Resort Hotel            0         24               2017   
2877    Resort Hotel            0         24               2017   
2878    Resort Hotel            0         24               2017   
...              ...          ...        ...                ...   
112499    City Hotel            0         13               2019   
113046    City Hotel            0         13               2019   
113082    City Hotel            0         13               2019   
113627    City Hotel            0        210               2019   
116451    City Hotel            0        191               2019   

       arrival_date_month  arrival_date_week_number  \
2392             December                        42   
2697             December                        44   
2867              January                        45   
2877              January                        45   
2878              January                        45   
...                   ...                       ...   
112499               July                        21   
113046               July                        22   
113082               July                        22   
113627             August                        23   
116451          September                        29   

        arrival_date_day_of_month  stays_in_weekend_nights  \
2392                           11                        2   
2697                           26                        7   
2867                            3                        0   
2877                            3                        2   
2878                            3                        3   
...                           ...                      ...   
112499                         24                        0   
113046                         29                        1   
113082                         29                        1   
113627                          9                        0   
116451                         16                        2   

        stays_in_weekdays_nights  adults  children  babies       meal  \
2392                           0       2       0.0       0  Breakfast   
2697                          15       1       0.0       0  Breakfast   
2867                           3       2       0.0       0  Breakfast   
2877                          10       1       0.0       0  Breakfast   
2878                          10       2       0.0       0  Breakfast   
...                          ...     ...       ...     ...        ...   
112499                         1       2       0.0       0  Breakfast   
113046                         3       1       0.0       0  Breakfast   
113082                         3       2       0.0       0  Breakfast   
113627                         1       2       0.0       0  Breakfast   
116451                         2       2       0.0       0  Breakfast   

                              city market_segment distribution_channel  \
2392                 Kota Denpasar      Online TA                TA/TO   
2697               Kabupaten Garut      Corporate            Corporate   
2867                     Kota Batu      Corporate            Corporate   
2877                 Kota Denpasar      Corporate            Corporate   
2878    Kabupaten Kepulauan Seribu      Corporate            Corporate   
...                            ...            ...                  ...   
112499                 Kota Cimahi      Online TA                TA/TO   
113046  Kabupaten Kepulauan Seribu  Offline TA/TO                TA/TO   
113082  Kabupaten Kepulauan Seribu  Offline TA/TO                TA/TO   
113627            Kabupaten Bangka         Direct               Direct   
116451               Kota Denpasar      Online TA                TA/TO   

        is_repeated_guest  previous_ca

### Handling missing values

In [ ]:
#for childreen : deleted
df_clean.dropna(subset=['children'],inplace=True)
df_clean['children'].isnull().sum() #check the total missing value

0

since the number of missing values is small, the row is deleted.

In [ ]:
#for city : deleted
df_clean.dropna(subset=['city'],inplace=True)
df_clean['city'].isnull().sum() #check the total missing value

0

since the number of missing values is small, the row is deleted.

In [ ]:
#for agent and company : delete column
df_clean.drop(columns=['company','agent'], axis=1, inplace=True)

In addition to many missing values, the column is not included for analysis.

## Duplicate data

### Checking duplicate data

In [ ]:
#Total duplicate data
df_clean.duplicated().sum()

33258

In [ ]:
#Display duplicate data
df_clean.loc[df_clean.duplicated(),:]

hotel  is_canceled  lead_time  arrival_date_year  \
5       Resort Hotel            0         14               2017   
22      Resort Hotel            0         72               2017   
43      Resort Hotel            0         70               2017   
138     Resort Hotel            1          5               2017   
200     Resort Hotel            0          0               2017   
...              ...          ...        ...                ...   
119349    City Hotel            0        186               2019   
119352    City Hotel            0         63               2019   
119353    City Hotel            0         63               2019   
119354    City Hotel            0         63               2019   
119373    City Hotel            0        175               2019   

       arrival_date_month  arrival_date_week_number  \
5               September                        27   
22              September                        27   
43              September                        27   
138             September                        28   
200             September                        28   
...                   ...                       ...   
119349            October                        35   
119352            October                        35   
119353            October                        35   
119354            October                        35   
119373            October                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
5                               1                        0   
22                              1                        2   
43                              2                        2   
138                             5                        1   
200                             7                        0   
...                           ...                      ...   
119349                         31                        0   
119352                         31                        0   
119353                         31                        0   
119354                         31                        0   
119373                         31                        1   

        stays_in_weekdays_nights  adults  children  babies       meal  \
5                              2       2       0.0       0  Breakfast   
22                             4       2       0.0       0  Breakfast   
43                             3       2       0.0       0     Dinner   
138                            0       2       0.0       0  Breakfast   
200                            1       1       0.0       0  Breakfast   
...                          ...     ...       ...     ...        ...   
119349                         3       2       0.0       0  Breakfast   
119352                         3       3       0.0       0  Breakfast   
119353                         3       3       0.0       0  Breakfast   
119354                         3       3       0.0       0  Breakfast   
119373                         3       1       0.0       0  Breakfast   

                       city market_segment distribution_channel  \
5          Kabupaten Bangka      Online TA                TA/TO   
22            Kota Denpasar         Direct               Direct   
43      Kabupaten Tangerang         Direct               Direct   
138           Kota Denpasar      Online TA                TA/TO   
200        Kabupaten Bangka      Online TA                TA/TO   
...                     ...            ...                  ...   
119349    Kabupaten Bandung      Online TA                TA/TO   
119352        Kota Surabaya      Online TA                TA/TO   
119353        Kota Surabaya      Online TA                TA/TO   
119354        Kota Surabaya      Online TA                TA/TO   
119373   Kabupaten Belitung  Offline TA/TO                TA/TO   

        is_repeated_guest  previous_cancellations  \
5                       0                       0   
22              

### Handling duplicate data

In [ ]:
df_clean.drop_duplicates(keep='last', inplace=True) #Remove duplicate data, by retrieving the last data
df_clean.duplicated().sum() ##Total duplicate data

0

## Data suitability

In [ ]:
#calculate the sum of values
for col in df_clean:
    print(f'''Value count column {col}:''')
    print(df_clean[col].value_counts())
    print('___'*20, '\n')
    print()

Value count column hotel:
City Hotel      52549
Resort Hotel    33091
Name: hotel, dtype: int64
____________________________________________________________ 


Value count column is_canceled:
0    61947
1    23693
Name: is_canceled, dtype: int64
____________________________________________________________ 


Value count column lead_time:
0      5798
1      3133
2      1879
3      1679
4      1535
       ... 
400       1
605       1
424       1
626       1
463       1
Name: lead_time, Length: 479, dtype: int64
____________________________________________________________ 


Value count column arrival_date_year:
2018    41562
2019    31335
2017    12743
Name: arrival_date_year, dtype: int64
____________________________________________________________ 


Value count column arrival_date_month:
October      11121
September     9946
July          8227
June          7772
August        7644
May           7378
December      6668
November      6426
April         5977
February      5029
January   

In [ ]:
print(df_clean['meal'].value_counts()) #Calculating unique values
print(print('___'*20, '\n'))
print(df_clean['distribution_channel'].value_counts()) #Calculating unique values

Breakfast     66643
No Meal        9434
Dinner         8767
Undefined       450
Full Board      346
Name: meal, dtype: int64
____________________________________________________________ 

None
TA/TO        68009
Direct       12685
Corporate     4767
GDS            178
Undefined        1
Name: distribution_channel, dtype: int64


There are undefined values (meal, distribution_channel)

In [ ]:
df_clean['meal'] = df_clean['meal'].replace('Undefined' , ) #Replace according to the previous row
df_clean['distribution_channel'] = df_clean['distribution_channel'].replace('Undefined' , ) #Replace according to the previous row
print(df_clean['meal'].value_counts()) #Calculating unique values
print(print('___'*20, '\n'))
print(df_clean['distribution_channel'].value_counts()) #Calculating unique values

Breakfast     66989
No Meal        9439
Dinner         8846
Full Board      366
Name: meal, dtype: int64
____________________________________________________________ 

None
TA/TO        68009
Direct       12686
Corporate     4767
GDS            178
Name: distribution_channel, dtype: int64


In [ ]:
df_clean.drop(columns=['arrival_date_week_number', 'required_car_parking_spaces'], axis=1, inplace=True) #remove unrelated columns

delete columns that are not needed

In [ ]:
print("before pre-processing: ", df.shape) #view the number of rows and columns before cleaning
print("after pre-processing: ", df_clean.shape) #view the number of rows and columns after cleaning

before pre-processing:  (119390, 29)
after pre-processing:  (85640, 25)


In [ ]:
#if you want to save the cleaned dataframe then run the following syntax.
#df_clean.to_csv('hotel_booking_data_clean.csv', index=False)